source/code/plugins/VMInsightsEngine.rb

```diff
@@ -0,0 +1,313 @@
+# frozen_string_literal: true
+
+module VMInsights
+
+    require_relative 'VMInsightsDataCollector.rb'
+
+    class MetricsEngine
+
+        require 'json'
+        require 'thread'
+
+        def initialize
+            @thread = nil
+        end
+
+        def start(config, &cb)
+            raise ArgumentError, 'config is nil' if config.nil?
+            raise ArgumentError, "config is not kind of #{Configuration}" unless config.kind_of? Configuration
+            raise ArgumentError, 'proc required' if cb.nil?
+            raise RuntimeError, 'already started' unless @thread.nil?
+            @thread = PollingThread.new config.poll_interval, config.computer, config.log, config.data_collector, cb
+            nil
+        end
+
+        def stop
+            return if @thread.nil?
+            @thread = nil if @thread.stop
+        end
+
+        def running?
+            (! @thread.nil?) && @thread.status
+        end
+
+        class Configuration
+            def initialize(computer, log, data_collector)
+                raise ArgumentError unless log
+                raise ArgumentError, "#{data_collector.class.name}" unless data_collector.kind_of? IDataCollector
+                @poll_interval = 60 # seconds
+                @computer = computer
+                @data_collector = data_collector
+                @log = log
+            end
+
+            def poll_interval=(v)
+                raise ArgumentError unless (v.kind_of? Numeric) && (v.real?) && (v >= 1)
+                @poll_interval = v
+            end
+
+            attr_reader :poll_interval, :computer, :data_collector, :log
+
+        end # class Configuration
+
+    private
+
+        class PollingThread < Thread
+            def initialize(interval, computer, log, data_collector, cb)
+                @mutex = Mutex.new
+                @condvar = ConditionVariable.new
+                @run = true
+                @log = log
+                @data_collector = data_collector
+                @cb = cb
+                @saved_exception = SavedException.new
+                @saved_cpu_exception = SavedException.new
+                @cummulative_data = CummulativeData.new
+                super() {
+                    @cummulative_data.initialize_from_baseline @data_collector.baseline
+                    MetricTuple.computer computer
+                    begin
+                            @log.info "Starting polling loop at #{interval} second interval"
+                            while @run do
+                                    @mutex.synchronize {
+                                            @condvar.wait(@mutex, interval) if @run
+                                    }
+                                    yield_metrics_message() if @run
```

Comment:
How does the code account for the time it takes to generate the metrics? It seems that the loop will always wait for interval seconds between each data collection, which is probably now what we want. Ideally, we want to collect the metrics once per 60 seconds, which means that we may need to adjust the wait time, based on how long we've spent collecting them.

Reply:
Will change.

Reply:
Changes completed.